src/libcollections/btree/map.rs

```diff
@@ -724,6 +786,84 @@ impl<K: Ord, V> BTreeMap<K, V> {
             })
         }
     }
+
+    fn from_sorted_iter<I: Iterator<Item=(K, V)>>(&mut self, iter: I) {
+        let mut cur_node = last_leaf_edge(self.root.as_mut()).into_node();
+        // Iterate through all key-value pairs, pushing them into nodes at the right level.
+        for (key, value) in iter {
+            // Try to push key-value pair into the current leaf node.
+            if cur_node.len() < node::CAPACITY {
+                cur_node.push(key, value);
+            } else {
+                // No space left, go up and push there.
+                let mut open_node;
+                let mut test_node = cur_node.forget_type();
+                loop {
+                    match test_node.ascend() {
+                        Ok(parent) => {
+                            let parent = parent.into_node();
+                            if parent.len() < node::CAPACITY {
+                                // Found a node with space left, push here.
+                                open_node = parent;
+                                break;
+                            } else {
+                                // Go up again.
+                                test_node = parent.forget_type();
+                            }
+                        },
+                        Err(node) => {
+                            // We are at the top, create a new root node and push there.
+                            open_node = node.into_root_mut().push_level();
+                            break;
+                        },
+                    }
+                }
+
+                // Push key-value pair and new right subtree.
+                let tree_height = open_node.height() - 1;
+                let mut right_tree = node::Root::new_leaf();
+                for _ in 0..tree_height {
+                    right_tree.push_level();
+                }
+                open_node.push(key, value, right_tree);
+
+                // Go down to the right-most leaf again.
+                cur_node = last_leaf_edge(open_node.forget_type()).into_node();
+            }
+
+            self.length += 1;
+        }
+    }
+
+    fn fix_right_edge(&mut self) {
+        // Handle underfull nodes, start from the top.
+        let mut cur_node = self.root.as_mut();
+        while let Internal(internal) = cur_node.force() {
+            // Check if right-most child is underfull.
+            let right_child = internal.last_edge().descend();
+            if right_child.len() < node::CAPACITY / 2 {
+                // We need to steal.
+                let num_steals = node::CAPACITY/2 - right_child.len();
+                match right_child.ascend() {
```

Comment:
Instead of `descend`ing to `right_child`, `ascending` back to `parent` (which is the same thing as `internal.last_edge()`), and then descending again, it might be better to calculate `right_child.len()` with a `reborrow`, then do the final `descend` unconditionally after possibly stealing:

```rust
let right_child_len = internal.reborrow().last_edge().descend().len();
if right_child_len < node::CAPACITY / 2 {
    // ...
}
cur_node = internal.last_edge().descend();
```

I think this might also help LLVM generate slightly better code, as it doesn't understand that `.descend().ascend()` does nothing.